### DISCUSSION TITLE
3D Secure 2.0 Support

### DISCUSSION THREAD

**Discussion by BassemN** [OpenMage/magento-lts#2080] [Extensions & Modules]
Hello,

Based on this [article](https://community.magento.com/t5/Magento-DevBlog/3D-Secure-2-0-changes/ba-p/136460)
> 3DS 1.0 is supported. When and/or where use of 3DS 2.0 is required, Merchants will either need to replace PayPal with Braintree or upgrade to Magento 2.3.x.

Magento 1 will not support 3DS 2 for PayPal pro so will this repo support the 3DS 2?

Thanks

**Comment by rudracomputech** [comment]
Hello Team, 
Do we have a chance to get 3ds 2 braintree with openMage if yes then when ?
Thanks

**Comment by justinbeaty** [comment]
@rudracomputech I would try to use the Gene Braintree module.

If you use composer, I have a fork of the module with the latest Braintree SDK here: https://github.com/justinbeaty/module-gene-braintree

If you do not use composer, you could download the old version here: https://github.com/justinbeaty/module-gene-braintree/archive/e0793fb2861abde2084d285cd8bb7653ad3dd640.zip

I am not sure if the module's built-in javascript files supports 3DS v2. I do know that the PHP sdk at least supports it, because I have 3DS v2 working but using some custom javascript.

**Comment by rudracomputech** [comment]
Hi @justinbeaty  thankyou for your response. I tried with your module but no luck I am getting "Card payment has failed, missing token/nonce" in log. 
I am using 
<pre>

Magento Braintree Payments Version	3.1.3
You're up to date!
You're running the latest version of our module.

For questions or assistance with the module, please visit our Support Portal.
Braintree PHP SDK Version	3.23.0
</pre>

Thanks in advance

**Comment by justinbeaty** [comment]
@rudracomputech Okay so it looks like you are using the ZIP file (aka the last official release from Gene).

For your issue, maybe it could be incorrect configuration in the Magento Admin Panel? Look under Sales -> Payment Methods.

First, does it say "Valid Credentials - You're ready to accept payments via Braintree" ? Second, is the "Braintree Payments - Credit Card" method enabled?

Finally, the last thing I ask you to check is if there are any JS error on the frontend during checkout. It is possible that the JS api that Gene originally used no longer works. I would not know for certain because I am using a custom frontend where I import Braintree's latest web framework. In other words I do not use Magento's default checkout flow.

**Comment by justinbeaty** [comment]
@rudracomputech I did try and checkout and I get the error. I think this is maybe a problem with your one page checkout customizations. Are you using a 3rd party checkout module, or just customized version of default Magento's checkout? Can you temporarily switch to the default onepagecheckout module from stock OpenMage just to debug? 

To diagnose, I checked Chrome's network log and there is no call to Braintree's API. Normally what should happen is that on click of the "continue" button, the Braintree module should call Braintree's API to generate a nonce for the payment, but this does not seem to be happening.

**Comment by fballiano** [comment]
I'm using gene_braintree on a customer of mine, with firecheckout, and it's still working fine with creditcards+paypal, does this help you in any way or could you need more info?

**Comment by justinbeaty** [comment]
@fballiano It does look like Gene has support for firecheckout:

```
user@local:~/modules/module-gene-braintree-public$ grep -Ri firecheckout app/
app/design/frontend/base/default/template/gene/braintree/js/firecheckout.phtml:<!-- FIRECHECKOUT BRAINTREE SUPPORT -->
```

But, @rudracomputech is this the checkout module you're using?

Also, if you would like @fballiano this can maybe be moved to discussions instead of an issue. (I think you may be able to move it with your repo permissions, not sure).

**Reply by fballiano** [reply]
- yes it has specific support for firecheckout
- moved to discussion

**Reply by rudracomputech** [reply]
yes I am using "IWD Extensions One Page Checkout v 4.2.2" third party checkout modules.

**Reply by fballiano** [reply]
gene_braintree should support iwd too, as far as i know

**Reply by justinbeaty** [reply]
Yes, looks like it does:

```
user@local:~/modules/module-gene-braintree-public$ grep -R iwd app/
app/design/frontend/base/default/template/gene/braintree/js/iwd.phtml: * https://www.iwdagency.com/extensions/one-step-page-checkout.html
```

So, have you made any customizations / overrides to this module? @rudracomputech

**Reply by rudracomputech** [reply]
nope I uploaded as you sent. very strange I dont know why I am getting nonce issue, seems its not calling javascript method enableDisableNonce

**Reply by justinbeaty** [reply]
@rudracomputech I did mean any modifications to the IWD module. 

One other thing to check, can you run from your htdocs root:

```
ls app/design/frontend/*/*/template/gene/braintree/js/iwd.phtml
```

There should be one entry:

```
app/design/frontend/base/default/template/gene/braintree/js/iwd.phtml
```

I ask because in view source, I see on your page:

```
<script type="text/javascript">
vZeroIntegration.addMethods({
/**
* Capture the original submit function
*/
captureOriginalSubmitFn: function () {
this._originalSubmitFn = Review.prototype.save;
this._originalNextStep = Review.prototype.nextStep;
this._originalPaymentSaveFn = Payment.prototype.save;
return true;
},
/// etc 
```

But this does not look like the file from the Gene module. Maybe it's been overridden.

**Reply by rudracomputech** [reply]
@justinbeaty yes the file is present on the above map. please share your email let me send you the ftp credentials

**Reply by justinbeaty** [reply]
@rudracomputech I apologize that I do not have enough spare time to check in via FTP. 

My question isn't if the file is present at `app/design/frontend/base/default/template/gene/braintree/js/iwd.phtml`, but if there is possibly another `iwd.phtml` file somewhere else in your `app/design/frontend` directory. If so, then you should try and remove it.

Do you have SSH access? If so then that command will tell you, otherwise you must traverse the directory tree yourself to try and find it.

**Reply by rudracomputech** [reply]
yes I have SSH access and I checked via locate command and there is only one iwd.phtml file present which is under gene/braintree/js directory

**Reply by fballiano** [reply]
better use `find . -name ' iwd.phtml'` instead of locate, the locate's database could be not updated (probably not this case, but anyway)

**Reply by rudracomputech** [reply]
yes same result

**Reply by justinbeaty** [reply]
Okay, well unless I am missing something, if you view-source on your checkout page and CTRL+F for `vZeroIntegration.addMethods`, it does not look like the file here: https://github.com/justinbeaty/module-gene-braintree/blob/e0793fb2861abde2084d285cd8bb7653ad3dd640/app/design/frontend/base/default/template/gene/braintree/js/iwd.phtml

Are you 100% sure credit card method is enabled? Make sure to check various "Current Configuration Scope:" in admin config.

**Reply by rudracomputech** [reply]
ok I solved it by changing default.phtml file with iwd.phtml under template/gene/braintree/js directory and its seems working good. @justinbeaty thanks for your help I really appreciate for sharing your time and knowledge again Thank you so much. God bless you :+1:

**Reply by justinbeaty** [reply]
@rudracomputech That seems like an okay solution. It seems the problem is that somehow the `opc_index_index` layout handle is not being applied properly from here: https://github.com/justinbeaty/module-gene-braintree/blob/e0793fb2861abde2084d285cd8bb7653ad3dd640/app/design/frontend/base/default/layout/gene/braintree.xml#L100

I am very glad you were able to solve this. So the next question is then does the Braintree v.zero SDK included in the Gene module support 3DS v2?

**Reply by rudracomputech** [reply]
@justinbeaty it seems working good with 3DS v2 👍
